python/triton/runtime/specialize.py

```diff
@@ -0,0 +1,89 @@
+import triton
+import textwrap
+import inspect
+import types
+import re
+
+import triton.runtime
+
+
+def define_kernel(src, attrs=None, **extra_globals):
+    '''
+    Dynamically create a Triton function or kernel from a src string,
+    linking any symbols in the kernel to objects specified by extra_globals.
+    '''
+
+    # create templace function
+    def _empty_fn():
+        pass
+
+    gdict = dict(**(_empty_fn.__globals__))
+    gdict.update(extra_globals)
+    f = types.FunctionType(_empty_fn.__code__, gdict)
+
+    src = textwrap.dedent(src)
+    src = src[src.find('def '):]
+
+    stored_functions = []
+    function_name = src[4:].split('(')[0].strip()
+
+    exec_globals = gdict
+    exec_globals.update({"stored_functions": stored_functions})
+    exec(src + '\n\nstored_functions.append(' + function_name + ')\n', exec_globals)
+
+    f.__signature__ = inspect.signature(stored_functions[0])
+    f.__name__ = function_name
+    f.__doc__ = stored_functions[0].__doc__
+
+    if attrs is None:
+        attrs = dict()
+    f = triton.JITFunction(f, **attrs)
+    f._unsafe_update_src(src)
+    return f
+
+
+def specialize(fn, constants, name=None):
+    assert isinstance(fn, triton.runtime.jit.JITFunction)
+    if name is None:
+        name = f"{fn.__name__}"
+    # Get original source code
+    src = inspect.getsource(fn.fn)
+    src = textwrap.dedent(src)
+    lines = src.split("\n")
+    # Skip decorator and def line
+    def_idx = next(i for i, line in enumerate(lines) if line.strip().startswith("def"))
+    # separate header vs body LOC
+    header_end = def_idx
+    while not lines[header_end].rstrip().endswith(':'):
+        header_end += 1
+    body_lines = lines[header_end + 1:]
+    header_lines = lines[def_idx:header_end + 1]
+    # clean-up header
+    header_clean = [
+        l.split("#", 1)[0].strip()  # keep code, discard comment
+        for l in header_lines
+        if l.split("#", 1)[0].strip()  # skip blank‑after‑comment lines
+    ]
+    # decompose arguments
+    header_src = " ".join(header_clean)  # turn it into a single line
+    m = re.search(r"\((.*)\)\s*:", header_src)
+    if not m:
+        raise ValueError("Could not parse function header")
+    args_str = m.group(1)
+    args = [arg.strip() for arg in args_str.split(",") if arg.strip()]
+    non_const_args = [arg for arg in args if arg.split(":")[0].split("=")[0].strip() not in constants]
+    # add global symbols
+    extra_globals = {v.__name__: v for k, v in constants.items() if isinstance(v, triton.runtime.jit.JITFunction)}
+    extra_globals.update(fn.__globals__)
+    # build new source code and define kernel dynamically
+    new_signature = f"def {name}({', '.join(non_const_args)}):"
```

Comment:
Is the idea to build a new `JITFunction` to minimize launch overhead? Otherwise it would seem simpler to just add a wrapper `SpecializedJITFunction` which automatically adds the extra arguments into the call, similar to `functools.partial`.